File: Examples/ore-rest-api/src/main/java/com/ore/api/service/StorageService.java
package com.ore.api.service;

import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.prefs.BackingStoreException;

/**
 * Service for managing job directories and files.
 */
@Slf4j
@Service
public class StorageService {

    private final Path jobsBaseDir;

    public StorageService(@Value("${ore.jobs.base-dir}") String jobsBaseDir) {
        this.jobsBaseDir = Path.of(jobsBaseDir);
    }

    /**
     * Create job input and output directories.
     */
    public List<Path> createJobDirs(String jobId) throws IOException {
        Path inputDir = jobsBaseDir.resolve(jobId).resolve("input");
        Path outputDir = jobsBaseDir.resolve(jobId).resolve("output");

        Files.createDirectories(inputDir);
        Files.createDirectories(outputDir);

        log.info("Created job directories for {}", jobId);
        return List.of(inputDir, outputDir);
    }

    public Path getJobInputDir(String jobId) {
        return jobsBaseDir.resolve(jobId).resolve("input");
    }

    public Path getJobOutputDir(String jobId) {
        return jobsBaseDir.resolve(jobId).resolve("output");
    }

    /**
     * List all files in the job output directory.
     */
    public List<String> getOutputFiles(String jobId) {
        Path outputDir = getJobOutputDir(jobId);
        if (!Files.exists(outputDir)) {
            return List.of();
        }

        try {
            return Files.list(outputDir)
                .filter(Files::isRegularFile)
                .map(Path::getFileName)
                .map(Path::toString)
                .toList();
        } catch (IOException e) {
            log.warn("Failed to list output files for job {}: {}", jobId, e.getMessage());
            return List.of();
        }
    }

    /**
     * Get validated file path for a job.
     * Prevents path traversal attacks.
     */
    public Path getFilePath(String jobId, String filename) {
        validateFilename(filename);

        // Check output directory first, then input
        for (String base : List.of("output", "input")) {
            Path path = jobsBaseDir.resolve(jobId).resolve(base).resolve(filename);
            Path resolved = path.toAbsolutePath().normalize();
            Path allowedDir = jobsBaseDir.resolve(jobId).resolve(base).toAbsolutePath().normalize();

            if (resolved.startsWith(allowedDir) && Files.exists(path)) {
                return path;
            }
        }

        throw new IllegalArgumentException("File not found: " + filename);
    }

    /**
     * Read ORE log file content.
     */
    public String readLog(String jobId) {
        for (String base : List.of("output", "input")) {
            Path logPath = jobsBaseDir.resolve(jobId).resolve(base).resolve("log.txt");
            if (Files.exists(logPath)) {
                try {
                    return Files.readString(logPath);
                } catch (IOException e) {
                    log.warn("Failed to read log for job {}: {}", jobId, e.getMessage());
                }
            }
        }
        return "";
    }

    /**
     * Validate filename to prevent path traversal attacks.
     */
    private void validateFilename(String filename) {
        if (filename == null || filename.isBlank()) {
            throw new IllegalArgumentException("Filename cannot be empty");
        }
        if (filename.contains("..") || filename.contains("/") || filename.contains("\\")) {
            throw new IllegalArgumentException("Invalid filename: " + filename);
        }
        if (Path.of(filename).isAbsolute()) {
            throw new IllegalArgumentException("Absolute paths not allowed: " + filename);
        }
    }
}

File: Examples/ORE-RESTAPI/app/services/storage_service.py
import logging
from pathlib import Path

from app.config import settings

logger = logging.getLogger(__name__)


class StorageService:
    """Manages per-job directory structure and file access."""

    def __init__(self):
        self.jobs_base_dir = Path(settings.JOBS_BASE_DIR)

    def create_job_dirs(self, job_id: str) -> tuple[Path, Path]:
        """Create jobs/{job_id}/input and jobs/{job_id}/output directories."""
        input_dir = self.jobs_base_dir / job_id / "input"
        output_dir = self.jobs_base_dir / job_id / "output"
        input_dir.mkdir(parents=True, exist_ok=True)
        output_dir.mkdir(parents=True, exist_ok=True)
        logger.info("Created job directories: %s", job_id)
        return input_dir, output_dir

    def get_job_input_dir(self, job_id: str) -> Path:
        return self.jobs_base_dir / job_id / "input"

    def get_job_output_dir(self, job_id: str) -> Path:
        return self.jobs_base_dir / job_id / "output"

    def get_output_files(self, job_id: str) -> list[str]:
        """List all files in the output directory."""
        output_dir = self.get_job_output_dir(job_id)
        if not output_dir.exists():
            return []
        return [f.name for f in output_dir.iterdir() if f.is_file()]

    def get_file_path(self, job_id: str, filename: str) -> Path:
        """Get validated file path for a job output file.

        Raises ValueError if path traversal is detected.
        """
        self._validate_filename(filename)

        # Check input directory first, then output
        for base in ["output", "input"]:
            path = self.jobs_base_dir / job_id / base / filename
            resolved = path.resolve()
            job_dir = (self.jobs_base_dir / job_id / base).resolve()
            if not str(resolved).startswith(str(job_dir)):
                raise ValueError(f"Path traversal detected: {filename}")
            if path.exists():
                return path

        raise FileNotFoundError(f"File not found: {filename}")

    def read_log(self, job_id: str) -> str:
        """Read the ORE log file."""
        # Check output dir first, then input dir
        for base in ["output", "input"]:
            log_path = self.jobs_base_dir / job_id / base / "log.txt"
            if log_path.exists():
                return log_path.read_text(encoding="utf-8", errors="replace")
        return ""

    def _validate_filename(self, filename: str) -> None:
        """Prevent path traversal attacks."""
        if not filename:
            raise ValueError("Filename cannot be empty")
        if ".." in filename:
            raise ValueError(f"Path traversal detected: {filename}")
        if "/" in filename or "\\" in filename:
            raise ValueError(f"Invalid filename (contains path separator): {filename}")
        if Path(filename).is_absolute():
            raise ValueError(f"Absolute paths not allowed: {filename}")


storage_service = StorageService()
